scheduler/src/cook/mesos/optimizer.clj

```diff
@@ -19,122 +19,251 @@
-
-                     Parameters:
-                     queue -- Ordered list of jobs to run (see cook.mesos.schema job)
-                     running -- Set of tasks running (see cook.mesos.schema instance)
-                     available -- Set of offers outstanding
-                     host-infos -- Host infos from HostFeed
-
-                     Returns:
-                     A schedule: a map where the keys are milliseconds in the future
-                     and  the values are recommendations to take at that point in the future"))
-
-(defn create-dummy-host-feed
-  "Returns an instance of HostFeed which returns an empty list of hosts"
-  [_]
-  (log/info "Creating dummy host feed")
-  (reify HostFeed
-    (get-available-host-info [this]
-      [])))
+  (produce-schedule
+    [this queue running]
+    "Returns a schedule of what jobs to place on what machines
+     and what hosts to purchase at different time steps.
+     Conforms to the Schedule schema above
+
+     Parameters:
+     queue -- Ordered list of jobs to run (see cook.mesos.schema job)
+     running -- Set of tasks running (see cook.mesos.schema instance)
+     host-infos -- Host infos from HostFeed
+
+     Returns:
+     A schedule: a map where the keys are milliseconds in the future
+     and  the values are recommendations to take at that point in the future"))
 
 (defn create-dummy-optimizer
   "Returns an instance of Optimizer which returns an empty schedule"
   [_]
   (log/info "Creating dummy optimizer")
   (reify Optimizer
-    (produce-schedule [this queue running available host-infos]
-      {0 {:suggested-matches {}}})))
+    (produce-schedule [_ _ _] {0 {:suggested-matches []}})))
 
-(defn optimizer-cycle!
+(defn instance-runtime
+  "Returns the instance runtime."
+  [instance now]
+  (let [start (:instance/start-time instance)
+        end (:instance/end-time instance now)]
+    (- (.getTime end)
+       (.getTime start))))
+
+(defn calculate-expected-complete-time
+  "Returns the expected time remaining to job completion.
+   Simple model for if run time exceeds expected runtime: the more it exceeds expected, the longer we expect it to take.
+   TODO: get better model!"
+  [instance expected-duration now]
+  (->> (instance-runtime instance now)
+       (- expected-duration)
+       Math/abs))
+
+(defn job->batch-uuid
+  "Retrieves a batch UUID for the provided job."
+  [job]
+  (or (-> job :group/_job :group/uuid)
+      (:job/uuid job)))
+
+(defn waiting-job->optimizer-job
+  "Converts a waiting job to an optimizer job representation."
+  [job default-runtime]
+  (let [expected-duration (get job :job/expected-runtime default-runtime)
+        resources (util/job-ent->resources job)
+        batch-uuid (job->batch-uuid job)]
+    (merge {:batch batch-uuid
+            :count 1
+            :dur expected-duration
+            :expected_complete_time -1
+            :running_host_group -1 ; TODO write function to determine this
```

Comment:
Same note about TODOs: If this comment will stay in the patch, then we should have a corresponding issue and reference it here.